src/cursor-doc/paredit.ts

```diff
@@ -792,6 +792,64 @@ export async function close(
   }
 }
 
+function onlyWhitespaceOnSameLineAsCursor(cursor: LispTokenCursor) {
+  cursor = cursor.clone();
+  while (!cursor.atStart()) {
+    switch (cursor.getPrevToken().type) {
+      case 'eol':
+        return true;
+      case 'ws':
+        cursor.previous();
+        continue;
+      default:
+        return false;
+    }
+  }
+}
+
+function backspaceOnWhitespaceEdit(doc: EditableDocument, cursor: LispTokenCursor) {
+  const start = doc.selection.anchor;
+  let trimStart = -1;
+  let linesSkipped = 0;
+  loop: while (!cursor.atStart()) {
+    cursor.previous();
+    switch (cursor.getToken().type) {
+      case 'eol':
+        if (linesSkipped > 0) {
+          trimStart = cursor.offsetEnd + 1;
+          break loop;
+        } else {
+          linesSkipped += 1;
+        }
+        break;
+      case 'ws':
+        trimStart = cursor.offsetStart;
+        break;
+      default:
+        trimStart = cursor.offsetEnd;
+        break loop;
+    }
+  }
+
+  if (trimStart === -1) {
+    return;
+  }
+
+  let selectionStart = trimStart;
+  const modelEdits = [new ModelEdit('deleteRange', [trimStart, start - trimStart])];
+  const cursorTokenType = cursor.getToken().type;
+  const destLineIsAllWhitespace =
+    cursorTokenType === 'ws' && onlyWhitespaceOnSameLineAsCursor(cursor);
+  if (!destLineIsAllWhitespace && cursorTokenType !== 'open') {
+    modelEdits.push(new ModelEdit('insertString', [trimStart, ' ']));
+    selectionStart += 1;
+  }
+  return doc.model.edit(modelEdits, {
+    selection: new ModelEditSelection(selectionStart),
+    skipFormat: false,
```

Comment:
Shouldn't we skip format to match your desired behaviour with leaving a space before the cursor in case we don't have any text after?

Reply:
I didn't have any problems with the space. Formatting may be ideal to avoid, but then we'd need a way to insert the correct number of spaces on your destination line. For example if you press backspace in this form:

https://user-images.githubusercontent.com/607986/170801894-965fd2bb-3c29-48ed-a8da-f4b7717d9342.mp4

You see how it shifts all the way to the left, then gets indented properly? If I change to skipFormat: true, it does not get indented properly. Do you know of a way to determine the correct number of spaces to insert?

Reply:
Here is how it works with the current code, despite doing the reformatting:

https://user-images.githubusercontent.com/607986/170802715-ca0a07cf-935a-4c4e-bf8f-562f39d38d4b.mp4

To me that looks correct, at least for those cases.

Reply:
Ah, yes, we do want the indents in the first case. Right now we only have a mechanism for post-formatting, maybe we should consider doing post-indentation instead/as an alternative... (for later, of course). We do have a way to calculate the indents, in case you want to explore using it for this edit:

https://github.com/BetterThanTomorrow/calva/blob/dev/src/cursor-doc/indent.ts#L147

(I'm fine with formatting, but putting in the correct indents would be less flickery.)